joomla/joomla-cms
# [cypress] Fail in case of SMTP problems with a clear message

### Summary of Changes

Improved Cypress system test failure messages in case of problems with SMTP.

If the array `mail[]` is empty width length 0, the line
`    cy.wrap(mails).should('have.lengthOf', 1);`
does not cause the test to stop, in this case, as the Cypress assertion inside the `.then()` has an asynchronous nature.

The effect is that the test run continues and the following access to the mail fails with with a somehow incomprehensible message e.g.
`    TypeError: Cannot read properties of undefined (reading 'body')`

Replaced it with an `expect` statement, which immediately triggers an error, and added an individual message that the SNMP configuration may need to be checked. Added two more places.

Trouble with `snmp-tester` port configuration was seen mutiple times, e.g. on configuring docker or on Ubuntu you need to open firewall with `ufw allow 1025`. May also be related to #42515 and #42516. Was discussed with @laoneo to check SNMP at the moment the `snmp-tester` is initialized. But that's only half the battle, because the SNMP configuration must work on `Node` and in the PHP-code of the server. This is why an individual test error message is used with a reference to the SNMP settings. And overall, a test must fail immediately if a mail is expected and is not there.

### Testing Instructions

* before PR
  * run all system tests e.g. with `npm run cypress:run` to see they are running without errors
  * _hack_ `configuration.php` with an unused, non-LISTEN port e.g. $smtpport = '6789' 
  * run the five system test specs with `smtp-tester` usage:
`  npm run cypress:run -- --spec 'tests/System/integration/administrator/components/com_config/Application.cy.js,tests/System/integration/site/components/com_users/Remind.cy.js,tests/System/integration/site/components/com_privacy/Request.cy.js,tests/System/integration/site/components/com_users/Reset.cy.js,tests/System/integration/api/com_contact/Contacts.cy.js'`
  * see the _ugly_ errors `Cannot read properties of undefined`
* after PR
  * run the five system test specs again and see the _nice_ errors `... you may need to check the SMTP configuration`
  * run all system tests e.g. with `npm run cypress:run` (including the initial Installation step which overwrites `configuration.php` with correct `snmpport`) to see they are running without errors

### Actual result BEFORE applying this Pull Request

system tests fail in case of trouble with errors like:
```
  1) Test in frontend that the privacy request view
       can submit an information request of type export without a menu item:
     TypeError: Cannot read properties of undefined (reading 'sender')
      at Context.eval (webpack://joomla/./tests/System/integration/site/components/com_privacy/Request.cy.js:16:23)
```

### Expected result AFTER applying this Pull Request

system tests fail in case of trouble with errors like:
```
  1) Test in frontend that the privacy request view
       can submit an information request of type export without a menu item:
     AssertionError: There are not two mails, you may need to check the SMTP configuration: expected 0 to equal 2
      at Context.eval (webpack://joomla/./tests/System/integration/site/components/com_privacy/Request.cy.js:15:103)
```

### Link to documentations
Please select:


### Upmerge

* after this PR is merged, please upmerge in 5.1-dev, 5.2-dev and 6.0-dev branches

### What was already tested?

* the PR was exhausted tested with the following steps:
  1) clone dev branch
     * on Mac and Windows applied 3 files from PR 43466 (as not merged so far)
     * on 6.0-dev applied 1 file from PR 43400 (as not upmerged so far)
     * install and run full system tests npm run run:cypress w/o errors 111 specs / 436 tests
  2) 'hacked' $smtpport = '6789' as unused non-LISTEN port and run five system tests specs with smtp-tester usage
     npm run cypress:run -- --spec 'tests/System/integration/administrator/components/com_config/Application.cy.js,tests/System/integration/site/components/com_users/Remind.cy.js,tests/System/integration/site/components/com_privacy/Request.cy.js,tests/System/integration/site/components/com_users/Reset.cy.js,tests/System/integration/api/com_contact/Contacts.cy.js'
     11 failing from 22 tests / 5 specs
     to see the _ugly_ errors
  3) fetched this PR and run again the 5 specs to see the _nice_ errors 11 failing from 22 tests / 5 specs
  4) run full system tests npm run run:cypress w/o errors (this is doing Web installation step first and overwrites configuration.php with snmpport 1025)

* did these four steps on the following platforms:
  a) local with branch 4.4-dev on Intel macOS 14.4.1 / Apache / MySQL
  b) local with branch 4.4-dev on Windows 11 Pro / Laragon
  c) on docker based installation
     . with branch 4.4-dev 435 tests (w/o Installation step)
     . with branch 5.1-dev 440 tests (w/o Installation step)
     . with branch 5.2-dev 440 tests (w/o Installation step)
     . with branch 6.0-dev 440 tests (w/o Installation step)
     
 * ignored test failure `api/com_menus/SiteMenuItems.cy.js` - `can create a site menu item`
    `Save failed with the following error: Joomla\\Component\\Menus\\Administrator\\Table\\MenuTable::_getNode(1, id) failed."` for the moment, assuming it is a mutliple test run problem
 * ignored two more test failures for the moment on 6.0-dev as they were also failing before the PR
